src/Lean/Elab/Notation.lean

```diff
@@ -43,15 +44,70 @@ def expandNotationItemIntoPattern (stx : Syntax) : MacroM Syntax :=
   else
     Macro.throwUnsupported
 
+partial def removeNonSemanticParenthesis (stx : Syntax) : MacroM Syntax := do
+  match stx with
+  | `(()) => pure stx -- Unit is relevant
+  | `(($e, $es,*)) => -- pairs are relevant
+    let e ← removeNonSemanticParenthesis e
+    let es ← es.getElems.mapM removeNonSemanticParenthesis
+    `(($e, $es,*))
+  | `(($e : $type)) => -- explicit types are relevant
+    let e ← removeNonSemanticParenthesis e
+    let type ← removeNonSemanticParenthesis type
+    `(($e : $type))
+  | `(($e)) =>
+    if let some stx ← Term.expandCDot? e then -- cdot notation doesn't have a delab, expand it
+      removeNonSemanticParenthesis stx
+    else
+      removeNonSemanticParenthesis e -- this parenthesis pair is non semantic, get rid off it
+  | stx =>
+    match stx with
+    | .node info kind args => pure $ .node info kind (←args.mapM removeNonSemanticParenthesis)
+    | _ => pure stx
+
+partial def hasDuplicateAntiquot (stx : Array Syntax) : StateM NameSet Bool :=
```

Comment:
```suggestion
partial def hasDuplicateAntiquot (stxs : Array Syntax) : StateM NameSet Bool :=
```
or `args`, just not `stx`

Reply:
You could replace the recursion + state monad with a loop over `Syntax.topDown` (we probably want `firstChoiceOnly := true` even if we may not handle choice nodes correctly yet in other places) and a mutable variable.